src/handshakes/ik.ts

```diff
@@ -0,0 +1,172 @@
+      session.h = h;
+      session.cs1 = cs1;
+      session.cs2 = cs2;
+    } else if (session.mc.gtn(1)) {
+      if (session.i) {
+        if (!session.cs1) {
+          throw new Error("CS1 (cipher state) is not defined")
+        }
+
+        messageBuffer = this.writeMessageRegular(session.cs1, message);
+      } else {
+        if (!session.cs2) {
+          throw new Error("CS2 (cipher state) is not defined")
+        }
+
+        messageBuffer = this.writeMessageRegular(session.cs2, message);
+      }
+    } else {
+      throw new Error("Session invalid.")
+    }
+
+    session.mc = session.mc.add(new BN(1));
+    return messageBuffer;
+  }
+
+  public recvMessage(session: NoiseSession, message: MessageBuffer): bytes {
+    let plaintext: bytes;
+    if (session.mc.eqn(0)) {
+      plaintext = this.readMessageA(session.hs, message);
+    } else if (session.mc.eqn(1)) {
+      const { plaintext: pt, h, cs1, cs2 } = this.readMessageB(session.hs, message);
+      plaintext = pt;
+      session.h = h;
+      session.cs1 = cs1;
+      session.cs2 = cs2;
+    } else if (session.mc.gtn(1)) {
+      if (session.i) {
+        if (!session.cs2) {
+          throw new Error("CS1 (cipher state) is not defined")
+        }
+        plaintext = this.readMessageRegular(session.cs2, message);
+      } else {
+        if (!session.cs1) {
+          throw new Error("CS1 (cipher state) is not defined")
+        }
+        plaintext = this.readMessageRegular(session.cs1, message);
+      }
+    } else {
+      throw new Error("Session invalid.");
+    }
+
+    session.mc = session.mc.add(new BN(1));
+    return plaintext;
+  }
+
+  private writeMessageA(hs: HandshakeState, payload: bytes): MessageBuffer {
+    hs.e = generateKeypair();
+    const ne = hs.e.publicKey;
+    this.mixHash(hs.ss, ne);
+    this.mixKey(hs.ss, this.dh(hs.e.privateKey, hs.rs));
+    const spk = Buffer.from(hs.s.publicKey);
+    const ns = this.encryptAndHash(hs.ss, spk);
+
+    this.mixKey(hs.ss, this.dh(hs.s.privateKey, hs.rs));
+    const ciphertext = this.encryptAndHash(hs.ss, payload);
+
+    return { ne, ns, ciphertext };
+  }
+
+  private writeMessageB(hs: HandshakeState, payload: bytes) {
+    hs.e = generateKeypair();
+    const ne = hs.e.publicKey;
+    this.mixHash(hs.ss, ne);
+
+    this.mixKey(hs.ss, this.dh(hs.e.privateKey, hs.re));
+    this.mixKey(hs.ss, this.dh(hs.e.privateKey, hs.rs));
+    const ciphertext = this.encryptAndHash(hs.ss, payload);
+    const ns = this.createEmptyKey();
+    const messageBuffer: MessageBuffer = {ne, ns, ciphertext};
+    const { cs1, cs2 } = this.split(hs.ss);
+
+    return { messageBuffer, cs1, cs2, h: hs.ss.h }
+  }
+
+  private readMessageA(hs: HandshakeState, message: MessageBuffer): bytes {
+    if (x25519.publicKeyVerify(message.ne)) {
+      hs.re = message.ne;
+    }
+
+    this.mixHash(hs.ss, hs.re);
+    this.mixKey(hs.ss, this.dh(hs.s.privateKey, hs.re));
+    const ns = this.decryptAndHash(hs.ss, message.ns);
```

Comment:
You are propagating error to  whoever is using handshake? 

Asking because go code is using 'valid' flags and has some conditions based on it?!

Reply:
They have different interface that includes those valid flags. In JS, if invalid, it's only throwing errors so I guess more try/catch blocks should be involved to substitute it.

Reply:
Depends, if flag if everything is always required to be valid you can just allow propagation and whoever is using handshake should handle error and not upgrade connection (should dorp connection probably)

Reply:
This is an interesting case. I have no idea if those validity flags are just Go thing or are required for the connection handling. However, execution of this part is different in the current JS version where if this line returns invalid, it doesn't continue execution which is the opposite of Go's.